storage/innobase/buf/buf0buf.cc

```diff
@@ -2486,80 +2574,119 @@ buf_block_t* buf_pool_t::page_fix(const page_id_t id)
   {
     hash_lock.lock_shared();
     buf_page_t *b= page_hash.get(id, chain);
-    if (b)
+    if (b && !watch_is_sentinel(*b))
     {
-      uint32_t state= b->fix();
-      hash_lock.unlock_shared();
+      uint32_t state= b->fix() + 1;
       ut_ad(!b->in_zip_hash);
-      ut_ad(b->frame);
-      ut_ad(state >= buf_page_t::FREED);
-      if (state >= buf_page_t::READ_FIX && state < buf_page_t::WRITE_FIX)
+      hash_lock.unlock_shared();
+
+      if (UNIV_UNLIKELY(state < buf_page_t::UNFIXED))
+      {
+        ut_ad(state >= buf_page_t::FREED);
+        /* The page was marked as freed or corrupted. */
+      unfix_corrupted:
+        b->unfix();
+      corrupted:
+        if (err)
+          *err= DB_CORRUPTION;
+        return nullptr;
+      }
+
+      if ((state >= buf_page_t::READ_FIX && state < buf_page_t::WRITE_FIX) ||
+          (state >= buf_page_t::IBUF_EXIST && state < buf_page_t::REINIT))
       {
+        if (no_wait)
+        {
+        would_block:
+          b->unfix();
+          return reinterpret_cast<buf_block_t*>(-1);
+        }
+
+        if (UNIV_UNLIKELY(!b->frame))
+        {
+        wait_for_unzip:
+          b->unfix();
+          std::this_thread::sleep_for(std::chrono::microseconds(100));
+          continue;
+        }
         b->lock.s_lock();
         state= b->state();
         ut_ad(state < buf_page_t::READ_FIX || state >= buf_page_t::WRITE_FIX);
+
+        if (state >= buf_page_t::IBUF_EXIST && state < buf_page_t::REINIT &&
+            buf_page_ibuf_merge_try(reinterpret_cast<buf_block_t*>(b),
+                                    RW_S_LATCH, err))
+          goto unfix_corrupted;
+
         b->lock.s_unlock();
       }
-      if (UNIV_UNLIKELY(state < buf_page_t::UNFIXED))
+
+      if (UNIV_UNLIKELY(!b->frame))
       {
-        /* The page was marked as freed or corrupted. */
-        b->unfix();
-        b= nullptr;
+        if (b->lock.x_lock_try());
+        else if (no_wait)
+          goto would_block;
+        else
+          goto wait_for_unzip;
+
+        buf_block_t *block= unzip(b, chain);
+        if (!block)
+          goto corrupted;
+
+        b= &block->page;
+        state= b->state();
+
+        if (state >= buf_page_t::IBUF_EXIST && state < buf_page_t::REINIT &&
+            buf_page_ibuf_merge_try(block, RW_X_LATCH, err))
+          goto unfix_corrupted;
+
+        b->lock.x_unlock();
       }
+
       return reinterpret_cast<buf_block_t*>(b);
     }
 
     hash_lock.unlock_shared();
-    switch (buf_read_page(id, 0)) {
+    switch (dberr_t local_err= buf_read_page(id)) {
```

Comment:
I think it is possible that at this point some other thread gets ahead of us and start loading the page. The call would then end up waiting for S lock in buf_read_page_low even if no_wait = T is specified as input. Can you please confirm if this is right ?

Reply:
Good catch. We’d better play it safe and refuse to wait for a page read if `no_wait=true`.

Reply:
Addressed in 7f1fbc69fc54c2f175027493829e05900e254063.